content/content-tool/tool/src/java/org/sakaiproject/content/entityproviders/ContentEntityProvider.java

```diff
@@ -260,6 +278,151 @@ public String getHtmlForRef(EntityView view, Map<String, Object> params) throws
+
+				DiskFileItem fileItem = (DiskFileItem) params.get("upload");
+
+				String[] fileNameParts = fileItem.getName().split("\\.(?=[^.]+$)");
+
+				String basename = StringUtils.equals(uploadFolderName, contentHostingService.getStudentUploadFolderName()) ? currentUser.getDisplayId() + "_" + fileNameParts[0] : fileNameParts[0];
+
+				securityService.pushAdvisor(allowedAdvisor);
+
+				ContentResourceEdit resourceEdit = contentHostingService.addResource(
+						uploadsFolder.getId(),
+						basename,
+						fileNameParts[1],
+						ContentHostingService.MAXIMUM_ATTEMPTS_FOR_UNIQUENESS);
+
+				resourceEdit.setContent(fileItem.getInputStream());
+				resourceEdit.setContentType(fileItem.getContentType());
+				resourceEdit.setContentLength(fileItem.getSize());
+
+				contentHostingService.commitResource(resourceEdit, NotificationService.NOTI_NONE);
+				JSONObject success = new JSONObject();
+				success.put("uploaded", 1);
+				success.put("fileName", fileItem.getName());
+				success.put("url", resourceEdit.getUrl());
+
+				return success.toJSONString();
+			}
+
+		} catch (TypeException | InconsistentException | IdInvalidException | IdUnusedException e) {
+			log.error("Exception when trying to get the direct upload folder for site " + context + ".", e);
+			errorMessage = "Unable to get uploads folder.";
+		} catch (OverQuotaException e) {
+			log.error("Over quota exception for site " + context + ".", e);
+			errorMessage = "Resource quota exceeded.";
+		} catch (ServerOverloadException e) {
+			log.error("Server overload exception for site " + context + ".", e);
+			errorMessage = "Server is busy. Please try again.";
+		} catch (PermissionException e) {
+			log.error("Permission exception trying to direct upload.", e);
+			errorMessage = "You do not have permission to perform this action.";
+		} catch (IdLengthException e) {
+			log.error("Resource file name too long.", e);
+			errorMessage = "File name too long.";
+		} catch (IdUniquenessException e) {
+			log.error("Unable to find unique id.", e);
+			errorMessage = "A file with that name already exists.";
+		} catch (IOException e) {
+			log.error("IOException with direct upload.", e);
+			errorMessage = "An error occurred trying to upload the file.";
+		} finally {
+			securityService.popAdvisor(allowedAdvisor);
+		}
+
+		// There was an error.
+		JSONObject errorMsg = new JSONObject();
+		errorMsg.put("message", errorMessage);
+		JSONObject error = new JSONObject();
+		error.put("uploaded", 0);
+		error.put("error", errorMsg);
+		return error.toJSONString();
+	}
+
+	private ContentCollection getDirectUploadsFolder(final String collectionId, final String uploadFolderName) throws TypeException, InconsistentException, IdInvalidException {
+
+		ContentCollection returnCollection = null;
+
+		try {
+			securityService.pushAdvisor(allowedAdvisor);
+
+			returnCollection = contentHostingService.getCollection(collectionId + uploadFolderName + "/");
+
+		} catch (IdUnusedException e) {
```

Comment:
I don't reckon you need all these blank lines here. Getting a bit overzealous with the readability :)